File: app/src/main/java/me/tagavari/airmessage/helper/ConversationColorHelper.java
package me.tagavari.airmessage.helper;

import android.util.SparseIntArray;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.List;
import java.util.Random;
import java.util.stream.Collectors;
import java.util.stream.IntStream;

import me.tagavari.airmessage.messaging.ConversationInfo;
import me.tagavari.airmessage.messaging.MemberInfo;

public class ConversationColorHelper {
	public static final Integer[] standardUserColors = {
			0xFFFF1744, //Red
			0xFFF50057, //Pink
			0xFFB317CF, //Purple
			0xFF703BE3, //Dark purple
			0xFF3D5AFE, //Indigo
			0xFF2979FF, //Blue
			0xFF00B0FF, //Light blue
			0xFF00B8D4, //Cyan
			0xFF00BFA5, //Teal
			0xFF00C853, //Green
			0xFF5DD016, //Light green
			0xFF99CC00, //Lime green
			0xFFF2CC0D, //Yellow
			0xFFFFC400, //Amber
			0xFFFF9100, //Orange
			0xFFFF3D00, //Deep orange
			//0xFF795548, //Brown
			//0xFF607D8B, //Blue grey
	};
	public static final int backupUserColor = 0xFF607D8B; //Blue grey
	
	private static Random getConversationRandom(ConversationInfo conversationInfo) {
		if(conversationInfo.getGUID() != null) return new Random(conversationInfo.getGUID().hashCode());
		else if(conversationInfo.getExternalID() != -1) return new Random(conversationInfo.getExternalID());
		else return new Random();
	}
	
	public static int getDefaultConversationColor(String conversationGUID) {
		return standardUserColors[Math.abs(conversationGUID.hashCode()) % standardUserColors.length];
	}
	
	public static int getDefaultConversationColor(long conversationExternalID) {
		return standardUserColors[(int) (Math.abs(conversationExternalID) % standardUserColors.length)];
	}
	
	public static int getDefaultConversationColor() {
		return standardUserColors[new Random().nextInt(standardUserColors.length)];
	}
	
	public static int getNextUserColor(ConversationInfo conversation) {
		//Creating a list of the user colors
		SparseIntArray colorUses = new SparseIntArray();
		
		//Adding all of the standard colors
		for(int color : standardUserColors) colorUses.put(color, 0);
		
		//Counting the colors
		for(MemberInfo member : conversation.getMembers()) {
			//Only allowing standard colors to be counted
			if(Arrays.stream(standardUserColors).noneMatch(standardColor -> standardColor == member.getColor())) continue;
			
			//Increasing the usage count
			colorUses.put(member.getColor(), colorUses.get(member.getColor(), 0) + 1);
		}
		
		//Finding the smallest use value
		int leastUses = conversation.getMembers().size();
		for(int i = 0; i < colorUses.size(); i++) {
			int uses = colorUses.valueAt(i);
			if(uses >= leastUses) continue;
			leastUses = uses;
			if(leastUses == 0) break;
		}
		
		//Finding all values with the least amount of uses
		ArrayList<Integer> leastUsedColors = new ArrayList<>();
		for(int i = 0; i < colorUses.size(); i++) {
			int uses = colorUses.valueAt(i);
			if(uses != leastUses) continue;
			leastUsedColors.add(colorUses.keyAt(i));
		}
		
		//Picking a least used color
		return leastUsedColors.get(getConversationRandom(conversation).nextInt(leastUsedColors.size()));
	}
	
	public static List<MemberInfo> getColoredMembers(String[] membersArray, int conversationColor) {
		return getColoredMembers(membersArray, conversationColor, new Random());
	}
	
	public static List<MemberInfo> getColoredMembers(String[] membersArray, int conversationColor, String conversationGUID) {
		return getColoredMembers(membersArray, conversationColor, new Random(conversationGUID.hashCode()));
	}
	
	public static List<MemberInfo> getColoredMembers(String[] membersArray, int conversationColor, long conversationExternalID) {
		return getColoredMembers(membersArray, conversationColor, new Random(conversationExternalID));
	}
	
	public static List<MemberInfo> getColoredMembers(String[] membersArray, int conversationColor, Random random) {
		//Inheriting the conversation color if there is only one member
		if(membersArray.length == 1) {
			return new ArrayList<>(Collections.singletonList(new MemberInfo(membersArray[0], conversationColor)));
		} else {
			//Sorting the members
			String[] members = membersArray.clone();
			Arrays.sort(members);
			
			//Getting color values
			int[] colorValues = ConversationColorHelper.getMassUserColors(random, membersArray.length);
			
			//Assigning the color values to the members
			return IntStream.range(0, members.length).mapToObj(i -> new MemberInfo(members[i], colorValues[i])).collect(Collectors.toCollection(ArrayList::new));
		}
	}
	
	private static int[] getMassUserColors(Random random, int userCount) {
		//Creating the color array
		int[] array = new int[userCount];
		
		//Adding the colors
		ArrayList<Integer> colors = new ArrayList<>();
		for(int i = 0; i < userCount; i++) {
			//Getting the colors if there are no more
			if(colors.isEmpty()) colors.addAll(Arrays.asList(standardUserColors));
			
			//Picking a color
			Integer color = colors.get(random.nextInt(colors.size()));
			
			//Setting the color
			array[i] = color;
			
			//Removing the color from use
			colors.remove(color);
		}
		
		//Returning the color array
		return array;
	}
}

File: app/src/main/java/me/tagavari/airmessage/helper/ConversationColorHelper.kt
package me.tagavari.airmessage.helper

import me.tagavari.airmessage.messaging.ConversationInfo
import me.tagavari.airmessage.messaging.MemberInfo
import kotlin.math.abs
import kotlin.random.Random

object ConversationColorHelper {
	@JvmField
	val standardUserColors: Array<Int> = arrayOf(
			0xFFFF1744.toInt(), //Red
			0xFFF50057.toInt(), //Pink
			0xFFB317CF.toInt(), //Purple
			0xFF703BE3.toInt(), //Dark purple
			0xFF3D5AFE.toInt(), //Indigo
			0xFF2979FF.toInt(), //Blue
			0xFF00B0FF.toInt(), //Light blue
			0xFF00B8D4.toInt(), //Cyan
			0xFF00BFA5.toInt(), //Teal
			0xFF00C853.toInt(), //Green
			0xFF5DD016.toInt(), //Light green
			0xFF99CC00.toInt(), //Lime green
			0xFFF2CC0D.toInt(), //Yellow
			0xFFFFC400.toInt(), //Amber
			0xFFFF9100.toInt(), //Orange
			0xFFFF3D00.toInt(), //Deep orange
			//0xFF795548.toInt(), //Brown
			//0xFF607D8B.toInt(), //Blue grey
	)
	const val backupUserColor = 0xFF607D8B.toInt() //Blue grey
	private fun getConversationRandom(conversationInfo: ConversationInfo): Random {
		return when {
			conversationInfo.guid != null -> Random(conversationInfo.guid!!.hashCode().toLong())
			conversationInfo.externalID != -1L -> Random(conversationInfo.externalID)
			else -> Random.Default
		}
	}
	
	@JvmStatic
	fun getDefaultConversationColor(conversationGUID: String): Int {
		return standardUserColors[abs(conversationGUID.hashCode()) % standardUserColors.size]
	}
	
	@JvmStatic
	fun getDefaultConversationColor(conversationExternalID: Long): Int {
		return standardUserColors[(abs(conversationExternalID) % standardUserColors.size).toInt()]
	}
	
	@JvmStatic
	fun getDefaultConversationColor(): Int {
		return standardUserColors[Random.nextInt(standardUserColors.size)]
	}
	
	@JvmStatic
	fun getNextUserColor(conversation: ConversationInfo): Int {
		val availableColors = conversation.members
				.map { it.color }
				//Only allowing standard colors to be counted
				.filter { memberColor -> standardUserColors.any { standardColor -> standardColor == memberColor } }
				//Group and count color usages
				.groupingBy { it }
				.eachCount()
				.entries
				//Filter to only keep results that have the lowest usage count
				.run {
					val leastUsedCount = fold(Int.MAX_VALUE) {acc, entry -> if(entry.value < acc) entry.value else acc}
					filter { entry -> entry.value == leastUsedCount }
				}
				//Map each entry to its color code value
				.map { it.key }
				//Default to standard user colors if there are no values
				.let { if(it.isEmpty()) standardUserColors.toList() else it }
		
		//Picking a least used color
		return availableColors[getConversationRandom(conversation).nextInt(availableColors.size)]
	}
	
	@JvmStatic
	fun getColoredMembers(members: List<String>, conversationColor: Int): MutableList<MemberInfo> {
		return getColoredMembers(members, conversationColor, Random.Default)
	}
	
	@JvmStatic
	fun getColoredMembers(members: List<String>, conversationColor: Int, conversationGUID: String): MutableList<MemberInfo> {
		return getColoredMembers(members, conversationColor, Random(conversationGUID.hashCode().toLong()))
	}
	
	@JvmStatic
	fun getColoredMembers(members: List<String>, conversationColor: Int, conversationExternalID: Long): MutableList<MemberInfo> {
		return getColoredMembers(members, conversationColor, Random(conversationExternalID))
	}
	
	@JvmStatic
	fun getColoredMembers(members: List<String>, conversationColor: Int, random: Random): MutableList<MemberInfo> {
		//Inheriting the conversation color if there is only one member
		return if(members.size == 1) {
			mutableListOf(MemberInfo(members[0], conversationColor))
		} else {
			var palette: MutableList<Int> = mutableListOf()
			
			//Sort the members alphabetically
			members.sorted()
					.map { address ->
						//Restock the palette if it is empty
						if(palette.isEmpty()) palette = standardUserColors.toMutableList()
						
						//Pick a random color from the palette
						val color = palette.random(random)
						palette.remove(color)
						
						MemberInfo(address, color)
					}
					.toMutableList()
		}
	}
}